### PR TITLE
Follow up for ConcurrentHashJoin

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

TODO:
- [ ] do not create dispatched blocks, just pass selector to addJoinedBlock 

https://github.com/ClickHouse/ClickHouse/pull/36415

### CODE REVIEW

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 1]
```diff
@@ -43,36 +46,45 @@ ConcurrentHashJoin::ConcurrentHashJoin(ContextPtr context_, std::shared_ptr<Tabl
...
-                }
+                /// if current hash_join is already processed by another thread, skip it and try later
+                std::unique_lock<std::mutex> lock(hash_join->mutex, std::try_to_lock);
+                if (!lock.owns_lock())
+                    continue;
 
-                hash_join->mutex.unlock();
-                iter = pending_blocks.erase(iter);
+                if (!dispatched_block)
+                    continue;
```
Comment (antonio2368) [inline src/Interpreters/ConcurrentHashJoin.cpp:78]: You can do this check before trying to acquire the lock

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 2]
```diff
@@ -16,25 +16,28 @@
 #include <Parsers/IAST_fwd.h>
 #include <Parsers/parseQuery.h>
 #include <Common/Exception.h>
+#include <Common/WeakHash.h>
 #include <Common/typeid_cast.h>
+#include <base/scope_guard.h>
```
Comment (rschu1ze) [inline src/Interpreters/ConcurrentHashJoin.cpp:21]: Is that for std::unique_lock? That would be in <mutex>.

File: src/Interpreters/ConcurrentHashJoin.cpp (hunk 1 above)
Comment (vdimir) [reply src/Interpreters/ConcurrentHashJoin.cpp:78]: Seems that it's possible that  another thread will do `dispatched_block = {}` between check and acquiring mutex.

File: src/Interpreters/ConcurrentHashJoin.cpp [hunk 3]
```diff
@@ -161,30 +173,32 @@ std::shared_ptr<NotJoinedBlocks> ConcurrentHashJoin::getNonJoinedBlocks(
     throw Exception(ErrorCodes::LOGICAL_ERROR, "Invalid join type. join kind: {}, strictness: {}", table_join->kind(), table_join->strictness());
 }
 
-Blocks ConcurrentHashJoin::dispatchBlock(const Strings & key_columns_names, const Block & from_block)
+static IColumn::Selector hashToSelector(const WeakHash32 & hash, size_t num_shards)
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Interpreters/ConcurrentHashJoin.cpp:176]: Usually I use anonymous namespace, except cases when function is used just once in a file. In that case I suppose it's good to have it near usage point to see its code without jumping in file

**Review by rschu1ze — APPROVED** [review]
The only concern (based on my superfluous understanding of the hash join code) is that method ::addJoinedBlock() contains a comment "if current hash_join is already processed by another thread, skip it and try later" - but since the code only tries to lock but does not wait for the lock - is it guaranteed that all items are processed in all situations? Would be bad to miss some elements, I guess.

**Comment by vdimir** [comment]
@rschu1ze

> is it guaranteed that all items are processed in all situations

Since we have loop while there're any unprocessed blocks, we will insert then eventually, when other thread will finish. Maybe it's good to fallback to lock with waiting after several steps to not perform unnecessary iterations, but for now decided to try to keep it simpler.

https://github.com/ClickHouse/ClickHouse/blob/4065b83ba99c6b394e5b4d9ec7fa905d420c795a/src/Interpreters/ConcurrentHashJoin.cpp#L62-L65

**Pull request merged**